### ISSUE TITLE
About the result

### ISSUE THREAD

**Issue by yao0801** [ChaojianYu/Hierarchical-Bilinear-Pooling#3]
@ChaojianYu sorry to trouble but I can't get the accuracy you have mentioned in your paper just through this project on the CUB dataset.
So I wonder if you use some tricks during your training, could you please tell me something interesting?
Thanks a lot.

**Comment by linrongc** [comment]
Just check the source code. It just linear project two vectors and element-wise product them together. 

The problem is that they use 1-dimension U (1xC) and V(1xV) to approximate W (CxC). 
W = U^T V
That bilinear pooling just downgrades to linear projection. I also can't reproduce the results and am just doubting the correctness of the reported accuracy.
Check the Factorized Bilinear Pooling Here:
http://winsty.net/papers/fbn.pdf

You will find the difference.

**Comment by ChaojianYu** [comment]
@yao0801 @linrongc  Thank you very much, this is a good question.  First of all, someone have already reproduced the results in the paper as far as I know. Some people can‘t get the corresponding results because the dataset is not processed properly. You can download the dataset I used here
[https://drive.google.com/drive/folders/17mhhfJnmtbTiDt5s74Mw9ftkEUlAHnn3](url)
I hope you can try it again.
If you have questions about the dataset provided. Below is the code for dataset preprocessing. You can reproduce it yourself.
[https://drive.google.com/file/d/1BKrjLrcHmJHtSuUwDInL5uaozikJIaKD/view?usp=sharing](url)
If you still can't reproduce the results using our code, here are the records of my experimental result, I hope this can help you.
[ft_last_layer.txt](https://github.com/ChaojianYu/Hierarchical-Bilinear-Pooling/files/2471401/ft_last_layer.txt)
[ft_all.txt](https://github.com/ChaojianYu/Hierarchical-Bilinear-Pooling/files/2471404/ft_all.txt)

 By the way, you can also share more about your experimental setup with me, and I am very happy to help you reproduce the results.

**Comment by ChaojianYu** [comment]
@linrongc FBP has fewer learning parameters than the original bilinear, and it also comes at the cost of training more epochs to converge.  You will find this If you compare it to the original bilinear.

In my opinion, I think the more important about the bilinear is that it can extract the correlation between features. Rather than the way it is implemented or is it mathematically equivalent.

**Comment by linrongc** [comment]
@ChaojianYu Thanks for the reply.

The provided links are not accessible.

In the original paper of FBN, the factor size they used is 20, which utilize  U(20xC), V(20xC) to approximate W because of the redundancy in W (low-rank in assumption). The accuracy they provided for CUB dataset is about 83%, which is close to my re-implementation. (Check Table 7 in the paper). 

If you only use a factor of 1, that means you only use 1/20 of the parameters and improve the accuracy to 85.4%(Fig 2).

That is why I am so confused and doubting the correctness of the reported accuracy.

**Comment by ChaojianYu** [comment]
@linrongc I am sorry for the inconvenience caused. Please open a browser window manually and copy and paste the entire link into the url address box.

FBP is a baseline in the paper. If you can't reproduce the result of FBP in our paper using the code, I can help you with it, and then for the performance difference you mentioned, I think there may be many factors in it. I suggest that you run our code first and then do further analysis.

**Comment by linrongc** [comment]
@ChaojianYu 
After some investigation into the problem, the performance difference mainly comes from the difference of the pre-trained models. 

Vgg16 in caffe is much better than that in pytorch, when it is used for transfer learning. By converting the caffe vgg16 weights to pytorch, I can now reproduce the results. 

For a fair comparison, I think you should reproduce other models under the same environments.
Using the same pretrained weight from caffe, now the basic Bilinear Pooling can boost the performance from 84% to 86%, with the same settings.

**Comment by daidaren202** [comment]
@linrongc Hello. I have the same question and doubt as you. I can't reproduce the result reported in the paper and get a accuracy closely as yours. And I alse doubt that  the baseline accuracy of this paper (85.7,FBP in table2) is so higher than original bilinear(84.1).  

Could you please share me your subsequent progress ? Thanks a lot.

**Comment by linrongc** [comment]
@daidaren202 the trick here is that. The vgg16 weight used in caffe is not the same with other platform. The pretrain process is much complex for the original vgg network. I tried to convert the original weight from caffe to pytorch and get much higher score for the model. But it seems the improvement is not restricted to the HBP model. Basic bilinear also perform much better than the original paper if you use the original vgg weight.

**Comment by daidaren202** [comment]
@linrongc Thanks for your reply firstly. And could you please share me the weight of vgg16 from caffe to pytorch as i am not fimiliar with caffe framework. I will be very grateful for your help!

**Comment by YuqiHUO** [comment]
I have reimplemented the experiment with Caffe, the results are true, although i couldn't reproduce the same result using pytorch...
